File: sdk-py/runrgateway/utils/retry.py
"""
Retry utilities with exponential backoff.
"""

import asyncio
import random
import time
from typing import Any, Callable, Optional, TypeVar

from ..errors import GatewayError, GatewayRateLimitError, GatewayUpstreamError

T = TypeVar('T')


class RetryOptions:
    """Configuration for retry behavior."""

    def __init__(
        self,
        max_retries: int = 3,
        base_delay: float = 1.0,
        max_delay: float = 10.0,
        jitter: bool = True,
    ):
        self.max_retries = max_retries
        self.base_delay = base_delay
        self.max_delay = max_delay
        self.jitter = jitter


def is_retryable_error(error: GatewayError) -> bool:
    """Check if an error is retryable."""
    # Rate limit errors are not retryable (wait for retry-after)
    if isinstance(error, GatewayRateLimitError):
        return False

    # Network errors are retryable
    if error.code == "NETWORK_ERROR":
        return True

    # Upstream errors (502, 503, 504) are retryable
    if isinstance(error, GatewayUpstreamError):
        return True

    # 5xx errors are retryable
    if error.status_code and error.status_code >= 500:
        return True

    return False


def calculate_delay(attempt: int, options: RetryOptions) -> float:
    """Calculate delay with exponential backoff and optional jitter."""
    delay = min(options.base_delay * (2 ** attempt), options.max_delay)

    if options.jitter:
        # Add ±25% jitter
        jitter = delay * 0.25 * (random.random() - 0.5)
        delay = max(0, delay + jitter)

    return delay


async def with_retry(
    fn: Callable[[], Any],
    options: Optional[RetryOptions] = None,
) -> T:
    """Retry a function with exponential backoff."""
    if options is None:
        options = RetryOptions()

    last_error: Exception

    for attempt in range(options.max_retries + 1):
        try:
            result = fn()
            if asyncio.iscoroutine(result):
                return await result
            return result
        except Exception as error:
            last_error = error

            # Don't retry on last attempt
            if attempt == options.max_retries:
                raise error

            # Check if error is retryable
            if isinstance(error, GatewayError) and not is_retryable_error(error):
                raise error

            # Calculate delay and wait
            delay = calculate_delay(attempt, options)
            await asyncio.sleep(delay)

    raise last_error

File: sdk-js/src/utils/retry.ts
import { GatewayError, GatewayRateLimitError, GatewayUpstreamError } from '../errors'

export interface RetryOptions {
  maxRetries: number
  baseDelay: number
  maxDelay: number
  jitter: boolean
}

export const DEFAULT_RETRY_OPTIONS: RetryOptions = {
  maxRetries: 3,
  baseDelay: 1000,
  maxDelay: 10000,
  jitter: true
}

/**
 * Check if an error is retryable
 */
export function isRetryableError(error: GatewayError): boolean {
  // Rate limit errors are not retryable (wait for retry-after)
  if (error instanceof GatewayRateLimitError) {
    return false
  }
  
  // Network errors are retryable
  if (error.code === 'NETWORK_ERROR') {
    return true
  }
  
  // Upstream errors (502, 503, 504) are retryable
  if (error instanceof GatewayUpstreamError) {
    return true
  }
  
  // 5xx errors are retryable
  if (error.statusCode && error.statusCode >= 500) {
    return true
  }
  
  return false
}

/**
 * Calculate delay with exponential backoff and optional jitter
 */
export function calculateDelay(
  attempt: number,
  options: RetryOptions
): number {
  const delay = Math.min(
    options.baseDelay * Math.pow(2, attempt),
    options.maxDelay
  )
  
  if (options.jitter) {
    // Add ±25% jitter
    const jitter = delay * 0.25 * (Math.random() - 0.5)
    return Math.max(0, delay + jitter)
  }
  
  return delay
}

/**
 * Sleep for a given number of milliseconds
 */
export function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms))
}

/**
 * Retry a function with exponential backoff
 */
export async function withRetry<T>(
  fn: () => Promise<T>,
  options: Partial<RetryOptions> = {}
): Promise<T> {
  const retryOptions = { ...DEFAULT_RETRY_OPTIONS, ...options }
  let lastError: Error
  
  for (let attempt = 0; attempt <= retryOptions.maxRetries; attempt++) {
    try {
      return await fn()
    } catch (error) {
      lastError = error as Error
      
      // Don't retry on last attempt
      if (attempt === retryOptions.maxRetries) {
        throw error
      }
      
      // Check if error is retryable
      if (error instanceof GatewayError && !isRetryableError(error)) {
        throw error
      }
      
      // Calculate delay and wait
      const delay = calculateDelay(attempt, retryOptions)
      await sleep(delay)
    }
  }
  
  throw lastError!
}
